toolchain/check/type_structure.cpp

```diff
@@ -387,15 +402,24 @@ class TypeStructureBuilder {
                type_id.has_value()) {
       Push(type_id);
     } else {
-      Push(NonTypeValue());
+      Push(NonTypeValue{.type_id = context_->insts().Get(inst_id).type_id()});
     }
   }
 
   // Push the next step into the work queue.
   auto Push(WorkItem item) -> void { work_list_.push_back(item); }
 
   // Append a structural element to the TypeStructure being built.
-  auto AppendStructural(TypeStructure::Structural structural) -> void {
+  auto AppendStructural(TypeStructure::Structural structural,
+                        TypeStructure::ConcreteType type =
+                            TypeStructure::ConcreteNoneType()) -> void {
```

Comment:
It seems like it's passed in specifically in a couple spots, and the default might only get used on for one `ConcreteCloseParen` call; maybe it could be made explicit there?

```suggestion
                        TypeStructure::ConcreteType type) -> void {
```

Or, perhaps go through the call sites that specify `ConcreteNoneType` and see if removing it feels clear enough?

Reply:
I think your comment above is a good one - splitting this into a few methods will resolve this.